### ISSUE TITLE
BigNum value reversed when retrieved from a key.

### ISSUE THREAD

**Issue by gh00000000** [openssl/openssl#21054]
If I load a key (ECDSA in the provided example) then ask for the private field right after, the bytes I get are different compared to the original.  In practice, the bytes are reversed.
It feels counter-intuitive, and it makes me doubt if I'm loading the key properly.

```
#include <openssl/ssl.h>
#include <openssl/engine.h>
#include <openssl/err.h>

#include <openssl/bn.h>
#include <openssl/evp.h>
#include <openssl/ec.h>
#include <openssl/core_names.h>
#include <stdint.h>

static const uint8_t ecdsa_private_key[] = {0x7d, 0x03, 0xeb, 0x52, 0x10, 0x60, 0x7f, 0xda, 0x87, 0x6e, 0xa0, 0xaa, 0xa4, 0xfc, 0x60, 0x25, 0xe6, 0x31, 0x5c, 0x7c, 0x6e, 0x70, 0x64, 0xeb, 0x76, 0xcf, 0x26, 0xec, 0xea, 0x64, 0xdc, 0x5e};

static const uint8_t ecdsa_public_key[] = { // X,Y
    POINT_CONVERSION_UNCOMPRESSED,
    0xb0, 0xc7, 0xa9, 0xb7, 0xe9, 0xd9, 0x5c, 0x42, 0x89, 0xb5, 0xd8, 0x1b, 0xa9, 0x12, 0x93, 0x49,
    0x23, 0x23, 0xf0, 0xef, 0xf5, 0xd6, 0x47, 0x6a, 0x72, 0x29, 0xce, 0x44, 0x7a, 0xd2, 0x36, 0x87,
    0x85, 0xe2, 0xcb, 0xf9, 0xea, 0x3c, 0x5b, 0x22, 0xce, 0x3f, 0x97, 0xf9, 0xa4, 0xda, 0x7d, 0xba,
    0xde, 0x97, 0xc9, 0xf4, 0x59, 0xd1, 0x48, 0xf3, 0x9b, 0xd2, 0xb3, 0xd7, 0xb5, 0x2d, 0xeb, 0x89
};

static void openssl_init()
{
    SSL_library_init();
    SSL_load_error_strings();
}

#define ECDSA_ID EVP_PKEY_EC
#define ECDSA_NID NID_X9_62_prime256v1
#define ECDSA_GROUP "P-256"

static EVP_PKEY *key_ecdsa_load(uint8_t *private_key, size_t private_key_size, uint8_t *public_key, size_t public_key_size)
{
    EVP_PKEY *evp_key = NULL;

    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(ECDSA_ID, NULL);
    if(pctx != NULL)
    {
        if(EVP_PKEY_paramgen_init(pctx) > 0)
        {
            if(EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, ECDSA_NID) > 0)
            {
                EVP_PKEY *params = NULL;

                if(EVP_PKEY_paramgen(pctx, &params) > 0)
                {
                    EVP_PKEY_CTX *kctx = EVP_PKEY_CTX_new(params, NULL);
                    if(kctx != NULL)
                    {
                        if(EVP_PKEY_fromdata_init(kctx) > 0)
                        {
                            OSSL_PARAM params[3] =
                            {
                                OSSL_PARAM_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME, ECDSA_GROUP, 0),
                                OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PUB_KEY, public_key, public_key_size),
                                OSSL_PARAM_BN(OSSL_PKEY_PARAM_PRIV_KEY, private_key, private_key_size),
                                OSSL_PARAM_END
                            };

                            if(EVP_PKEY_fromdata(kctx, &evp_key, EVP_PKEY_KEYPAIR, params) > 0)
                            {
                                uint8_t private_key_retrieved[128];
                                BIGNUM *priv = NULL;
                                if(EVP_PKEY_get_bn_param(evp_key, OSSL_PKEY_PARAM_PRIV_KEY, &priv) > 0)
                                {
                                    size_t private_key_retrieved_size = BN_bn2bin(priv, private_key_retrieved);
                                    if(private_key_size == private_key_retrieved_size)
                                    {
                                        if(memcmp(private_key, private_key_retrieved, private_key_size) != 0)
                                        {
                                            puts("The key bytes do not match."); // <- this line is hit
                                        }
                                        int reversed = 1;
                                        for(int i = 0; i < private_key_size; ++i)
                                        {
                                            if(private_key[i] != private_key_retrieved[private_key_size - 1 - i])
                                            {
                                                reversed = 0;
                                                break;
                                            }
                                        }
                                        if(reversed)
                                        {
                                            puts("Key bytes are reversed"); // <- this line is hit too
                                        }
                                    }
                                    else
                                    {
                                        puts("Size doesn't match.");
                                    }
                                }
                            }
                        }

                        EVP_PKEY_CTX_free(kctx);
                    }

                    EVP_PKEY_free(params);
                }
            }
        }
        EVP_PKEY_CTX_free(pctx);
    }

    return evp_key;
}

static void openssl_finalise()
{
    ERR_free_strings();
    EVP_cleanup();
    CRYPTO_cleanup_all_ex_data();
    ENGINE_cleanup();
}

int
main(int argc, char *argv[])
{
    (void)argc;
    (void)argv;

    openssl_init();

    key_ecdsa_load(ecdsa_private_key, sizeof(ecdsa_private_key), ecdsa_public_key, sizeof(ecdsa_public_key));

    openssl_finalise();

    return EXIT_SUCCESS;
}

```

**Comment by gh00000000** [comment]
note: OSSL_PARAM params[3] should be OSSL_PARAM params[4]

**Comment by t8m** [comment]
You should NOT use OSSL_PARAM_BN() unless you know what you're doing. The bignums in OSSL_PARAMs must be encoded with native endianness. I.e., little endian when on LE platforms, big endian on BE platforms.

On the other hand BN_bn2bin converts the BIGNUM always to big endian byte order. You're most likely on little endian platform and that's why you see the reversal.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input

**Comment by t8m** [comment]
The question was answered. Closing.
